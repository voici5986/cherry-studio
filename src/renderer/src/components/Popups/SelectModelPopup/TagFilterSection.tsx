import { loggerService } from '@logger'
import {
  EmbeddingTag,
  FreeTag,
  ReasoningTag,
  RerankerTag,
  ToolsCallingTag,
  VisionTag,
  WebSearchTag
} from '@renderer/components/Tags/Model'
import { ModelTag } from '@renderer/types'
import { Flex } from 'antd'
import React, { startTransition, useCallback, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

const logger = loggerService.withContext('TagFilterSection')

interface TagFilterSectionProps {
  availableTags: ModelTag[]
  tagSelection: Record<ModelTag, boolean>
  onToggleTag: (tag: ModelTag) => void
}

const TagFilterSection: React.FC<TagFilterSectionProps> = ({ availableTags, tagSelection, onToggleTag }) => {
  const { t } = useTranslation()

  const handleTagClick = useCallback(
    (tag: ModelTag) => {
      startTransition(() => onToggleTag(tag))
    },
    [onToggleTag]
  )

  // 标签组件
  const tagComponents = useMemo(
    () => ({
      vision: VisionTag,
      embedding: EmbeddingTag,
      reasoning: ReasoningTag,
      function_calling: ToolsCallingTag,
      web_search: WebSearchTag,
      rerank: RerankerTag,
      free: FreeTag
    }),
    []
  )

  return (
    <FilterContainer>
      <Flex wrap="wrap" gap={4}>
        <FilterText>{t('models.filter.by_tag')}</FilterText>
        {availableTags.map((tag) => {
          const TagElement = tagComponents[tag]
          if (!TagElement) {
            logger.error(`Tag element not found for tag: ${tag}`)
            return null
          }
          return (
            <TagElement
              key={`tag-${tag}`}
              onClick={() => handleTagClick(tag)}
              inactive={!tagSelection[tag]}
              showLabel
            />
          )
        })}
      </Flex>
    </FilterContainer>
  )
}

const FilterContainer = styled.div`
  padding: 8px;
  padding-left: 18px;
`

const FilterText = styled.span`
  color: var(--color-text-3);
  font-size: 12px;
`

export default TagFilterSection
