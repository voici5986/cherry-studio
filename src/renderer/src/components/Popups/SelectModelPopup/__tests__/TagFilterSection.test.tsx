import type { ModelTag } from '@renderer/types'
import { fireEvent, render, screen } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import TagFilterSection from '../TagFilterSection'

const mocks = vi.hoisted(() => ({
  t: vi.fn((key: string) => key),
  createTagComponent: (name: string) => {
    // Create a simple button component exposing props for assertions
    return ({ onClick, inactive, showLabel }: { onClick?: () => void; inactive?: boolean; showLabel?: boolean }) => {
      const React = require('react')
      return React.createElement(
        'button',
        {
          type: 'button',
          'aria-label': `tag-${name}`,
          'data-inactive': String(Boolean(inactive)),
          onClick
        },
        showLabel ? name : ''
      )
    }
  }
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => ({ t: mocks.t })
}))

vi.mock('@renderer/components/Tags/Model', () => ({
  VisionTag: mocks.createTagComponent('vision'),
  EmbeddingTag: mocks.createTagComponent('embedding'),
  ReasoningTag: mocks.createTagComponent('reasoning'),
  ToolsCallingTag: mocks.createTagComponent('function_calling'),
  WebSearchTag: mocks.createTagComponent('web_search'),
  RerankerTag: mocks.createTagComponent('rerank'),
  FreeTag: mocks.createTagComponent('free')
}))

vi.mock('antd', () => ({
  Flex: ({ children }: { children: React.ReactNode }) => children
}))

function createSelection(overrides: Partial<Record<ModelTag, boolean>> = {}): Record<ModelTag, boolean> {
  const base: Record<ModelTag, boolean> = {
    vision: true,
    embedding: true,
    reasoning: true,
    function_calling: true,
    web_search: true,
    rerank: true,
    free: true
  }
  return { ...base, ...overrides }
}

const allTags: ModelTag[] = ['vision', 'embedding', 'reasoning', 'function_calling', 'web_search', 'rerank', 'free']

describe('TagFilterSection', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <TagFilterSection availableTags={allTags} tagSelection={createSelection()} onToggleTag={vi.fn()} />
      )
      expect(container).toMatchSnapshot()
    })

    it('should reflect inactive state based on tagSelection', () => {
      render(
        <TagFilterSection
          availableTags={['vision']}
          tagSelection={createSelection({ vision: false })}
          onToggleTag={vi.fn()}
        />
      )
      const visionBtn = screen.getByRole('button', { name: 'tag-vision' })
      expect(visionBtn).toHaveAttribute('data-inactive', 'true')
    })

    it('should skip unknown tags', () => {
      render(
        <TagFilterSection
          availableTags={['unknown' as unknown as ModelTag, 'vision']}
          tagSelection={createSelection()}
          onToggleTag={vi.fn()}
        />
      )
      expect(screen.queryByRole('button', { name: 'tag-unknown' })).not.toBeInTheDocument()
      expect(screen.getByRole('button', { name: 'tag-vision' })).toBeInTheDocument()
    })
  })

  describe('functionality', () => {
    it('should call onToggleTag when a tag is clicked', () => {
      const handleToggle = vi.fn()
      render(<TagFilterSection availableTags={allTags} tagSelection={createSelection()} onToggleTag={handleToggle} />)

      const visionBtn = screen.getByRole('button', { name: 'tag-vision' })
      fireEvent.click(visionBtn)

      expect(handleToggle).toHaveBeenCalledTimes(1)
      expect(handleToggle).toHaveBeenCalledWith('vision')
    })
  })
})
