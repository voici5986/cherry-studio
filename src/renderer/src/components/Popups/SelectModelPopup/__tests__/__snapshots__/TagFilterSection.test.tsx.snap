// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TagFilterSection > rendering > should match snapshot 1`] = `
.c0 {
  padding: 8px;
  padding-left: 18px;
}

.c1 {
  color: var(--color-text-3);
  font-size: 12px;
}

<div>
  <div
    class="c0"
  >
    <span
      class="c1"
    >
      models.filter.by_tag
    </span>
    <button
      aria-label="tag-vision"
      data-inactive="false"
      type="button"
    >
      vision
    </button>
    <button
      aria-label="tag-embedding"
      data-inactive="false"
      type="button"
    >
      embedding
    </button>
    <button
      aria-label="tag-reasoning"
      data-inactive="false"
      type="button"
    >
      reasoning
    </button>
    <button
      aria-label="tag-function_calling"
      data-inactive="false"
      type="button"
    >
      function_calling
    </button>
    <button
      aria-label="tag-web_search"
      data-inactive="false"
      type="button"
    >
      web_search
    </button>
    <button
      aria-label="tag-rerank"
      data-inactive="false"
      type="button"
    >
      rerank
    </button>
    <button
      aria-label="tag-free"
      data-inactive="false"
      type="button"
    >
      free
    </button>
  </div>
</div>
`;
